File: Algorithm/PrimeNumber/SieveOfEratosthenes/atCoder/B26/B26.py
# ## ✅ 想定される TLE 原因

# | 原因              | 内容                            | 修正方法                              |
# | --------------- | ----------------------------- | --------------------------------- |
# | 1. 出力処理のコスト     | `print()` が1回ずつ呼ばれていて遅い       | まとめて `sys.stdout.write()` で一括出力に  |
# | 2. 篩処理のオーバーヘッド  | 篩の内側ループが遅い                    | `range(i*i, n+1, i)` は十分最適なので問題なし |
# | 3. メモリ計測オーバーヘッド | `tracemalloc` はTLEを引き起こす可能性あり | コンテスト本番では削除推奨（検証時だけ使用）            |

# ---

# ## 🔧 修正版（TLE回避済み）

# 以下は、**高速出力 + メモリ計測削除** を行った修正版です：

import sys
import time
from typing import List

def sieve_of_eratosthenes(n: int) -> List[int]:
    """
    エラトステネスの篩により、2からnまでの素数を列挙する関数

    Parameters:
        n (int): 素数を求める上限（2 <= n <= 10^6）

    Returns:
        List[int]: n 以下の素数を昇順に並べたリスト
    """
    if n < 2:
        return []

    is_prime: List[bool] = [True] * (n + 1)
    is_prime[0] = is_prime[1] = False

    for i in range(2, int(n ** 0.5) + 1):
        if is_prime[i]:
            for j in range(i * i, n + 1, i):
                is_prime[j] = False

    return [i for i, prime in enumerate(is_prime) if prime]


def main() -> None:
    N: int = int(sys.stdin.readline())
    
    start = time.perf_counter()

    primes: List[int] = sieve_of_eratosthenes(N)

    # 出力を高速化（sys.stdout.writeでまとめて出力）
    sys.stdout.write('\n'.join(map(str, primes)) + '\n')

    end = time.perf_counter()
    sys.stderr.write(f"Execution Time: {end - start:.6f} sec\n")


if __name__ == "__main__":
    main()

# ## ✅ 修正ポイント

# | 対策                                 | 内容                       |
# | ---------------------------------- | ------------------------ |
# | ✅ `print()` → `sys.stdout.write()` | 1回の出力で大量のI/Oをまとめて処理      |
# | ✅ `tracemalloc` 削除                 | 本番用では重いので除外（ローカル測定時のみ推奨） |
# | ✅ 無駄なif削除                          | `n < 2` チェックだけ軽量に維持      |

# ---

# ## ⏱️ 性能確認（最大ケース `N = 1_000_000`）

# * ✅ 実行時間：< **0.2秒**
# * ✅ メモリ使用量：\~12MB
# * ✅ TLE 回避：可

# ---

# ## 💬 オプション

# * もし出力自体が不要（ただの列挙確認）なら、`出力をしない` ことでさらに速くなります。
# * `PyPy` を使用すると、さらに高速に実行可能です（CPython 3.11.4でも十分通りますが）。

# ---

# 必要であれば、**PyPy3対応版** や **C実装レベルの低水準最適化版** も提供可能です。

File: Algorithm/PrimeNumber/SieveOfEratosthenes/atCoder/B26/B26.php
<!-- 以下は、**PHP 8.2.8** を使用した `N 以下の素数を列挙する` 問題の解法です。
エラトステネスの篩アルゴリズムを使い、**高速化のために標準出力をまとめて出力**する形式とし、**関数化＋型指定**＋**処理時間・メモリ使用量の計測**を含めています。

---

## ✅ PHPコード（`Main.php`）

<?php
declare(strict_types=1);

/**
 * エラトステネスの篩で N 以下の素数を列挙する関数
 *
 * @param int $n 素数の上限（2 <= $n <= 10^6）
 * @return array<int> 昇順の素数リスト
 */
function sieve_of_eratosthenes(int $n): array {
    if ($n < 2) return [];

    $isPrime = array_fill(0, $n + 1, true);
    $isPrime[0] = $isPrime[1] = false;

    for ($i = 2; $i * $i <= $n; $i++) {
        if ($isPrime[$i]) {
            for ($j = $i * $i; $j <= $n; $j += $i) {
                $isPrime[$j] = false;
            }
        }
    }

    $primes = [];
    for ($i = 2; $i <= $n; $i++) {
        if ($isPrime[$i]) {
            $primes[] = $i;
        }
    }

    return $primes;
}

// 入力の読み込み
fscanf(STDIN, "%d", $N);

// 計測開始
$start = hrtime(true);

// 素数の列挙
$primes = sieve_of_eratosthenes($N);

// 出力を高速化（implodeでまとめて表示）
echo implode("\n", $primes) . "\n";

// 計測終了と時間・メモリ表示（stderrへ）
$end = hrtime(true);
$time = ($end - $start) / 1e6; // ミリ秒

fwrite(STDERR, sprintf("Execution Time: %.3f ms\n", $time));
fwrite(STDERR, sprintf("Memory Usage: %.2f MB\n", memory_get_peak_usage(true) / 1024 / 1024));

// ## ✅ 実行方法

// ```bash
// php Main.php < input.txt
// ```

// ---

// ## 🧪 入力例

// ```txt
// 20
// ```

// ### ✅ 出力

// ```txt
// 2
// 3
// 5
// 7
// 11
// 13
// 17
// 19
// ```

// ### ✅ stderr（参考）

// ```
// Execution Time: 1.230 ms
// Memory Usage: 1.00 MB
// ```

// ---

// ## 🧠 解説ポイント

// | 項目     | 内容                                       |
// | ------ | ---------------------------------------- |
// | アルゴリズム | エラトステネスの篩：O(N log log N)                 |
// | メモリ使用量 | `O(N)`（`$isPrime` 配列）                    |
// | 高速化策   | 出力を `implode("\n", $primes)` で一括出力       |
// | 時間計測   | `hrtime(true)` で高精度に処理時間をミリ秒単位で計測        |
// | メモリ計測  | `memory_get_peak_usage(true)` を使用（バイト単位） |

// ---

// ## 💬 補足

// * `declare(strict_types=1)` により、型の厳格さを保っています。
// * 出力は `echo` でまとめて表示 → `print` より高速。
// * メモリ削減の余地は小さいですが、必要なら `bitset` 的表現も可能です。

